File: Java/OPSJLib/src/ops/Domain.java
/**
*
* Copyright (C) 2006-2009 Anton Gravestam.
*
* This file is part of OPS (Open Publish Subscribe).
*
* OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.

* OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
*/

package ops;

import configlib.ArchiverInOut;
import java.io.IOException;
import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.util.Enumeration;
import java.util.Vector;

/**
 *
 * @author angr
 */
public class Domain extends OPSObject
{
    private String domainAddress = "";
    private String domainID = "";
    private String localInterface = "0.0.0.0";
    protected Vector<Topic> topics = new Vector<Topic>();
    private int timeToLive = 1;
    private int inSocketBufferSize = 16000000;
    private int outSocketBufferSize = 16000000;
    private int metaDataMcPort = 9494;


    public Domain()
    {
        appendType("Domain");
    }
    

    public Topic getTopic(String name)
    {
        for (Topic topic : topics)
        {
            if(topic.getName().equals(name))
            {
                if(topic.getDomainAddress().equals(""))
                {
                    topic.setDomainAddress(domainAddress);
                }
                return topic;
            }
        }
        return null;
    }

    @Override
    public void serialize(ArchiverInOut archive) throws IOException
    {
        // NOTE. Keep this in sync with the C++ version, so it in theory is possible to send these as objects.
        // We need to serialize fields in the same order as C++.
        //OPSObject::serialize(archiver);
        super.serialize(archive);

        //archiver->inout(std::string("domainID"), domainID);
        //archiver->inout<Topic>(std::string("topics"), topics);
        //archiver->inout(std::string("domainAddress"), domainAddress);
        //archiver->inout(std::string("localInterface"), localInterface);
        domainID = archive.inout("domainID", domainID);
        topics = (Vector<Topic>) archive.inoutSerializableList("topics", topics);
        domainAddress = archive.inout("domainAddress", domainAddress);
        localInterface = archive.inout("localInterface", localInterface);

        //archiver->inout(std::string("timeToLive"), timeToLive);
        //archiver->inout(std::string("inSocketBufferSize"), inSocketBufferSize);
        //archiver->inout(std::string("outSocketBufferSize"), outSocketBufferSize);
        //archiver->inout(std::string("metaDataMcPort"), metaDataMcPort);
        timeToLive = archive.inout("timeToLive", timeToLive);
	inSocketBufferSize = archive.inout("inSocketBufferSize", inSocketBufferSize);
    	outSocketBufferSize = archive.inout("outSocketBufferSize", outSocketBufferSize);
        metaDataMcPort = archive.inout("metaDataMcPort", metaDataMcPort);
    }


    public Vector<Topic> getTopics() {
        for (Topic topic : topics) {
            if (topic.getDomainAddress().equals("")) {
                topic.setDomainAddress(domainAddress);
            }
        }
        return topics;
    }

     public boolean existTopic(String name) {
        for (Topic topic : topics) {
            if (topic.getName().equals(name)) {
                return true;
            }
        }
        return false;
    }

    public String getDomainAddress() {
        return domainAddress;
    }

    public String getDomainID() {
        return domainID;
    }

    public String getLocalInterface() {
        return localInterface;
    }

    public int GetMetaDataMcPort()
    {
        return metaDataMcPort;
    }

    public int GetInSocketBufferSize()
    {
        return inSocketBufferSize;
    }

    public int GetOutSocketBufferSize()
    {
        return outSocketBufferSize;
    }

    public int getTimeToLive()
    {
        return timeToLive;
    }

    public void setDomainAddress(String domainAddress) {
        this.domainAddress = domainAddress;
    }

    public void setDomainID(String domainID) {
        this.domainID = domainID;
    }

    public void setLocalInterface(String localInterface) {
        this.localInterface = localInterface;
    }

    // If argument contains a "/" we assume it is on the form:  subnet-address/subnet-mask
    // In that case we loop over all interfaces and take the first one that matches
    // i.e. the one whos interface address is on the subnet
    public static String DoSubnetTranslation(String ip)
    {
        int index = ip.indexOf('/');
        if (index < 0) return ip;

        String subnetIp = ip.substring(0, index);
        String subnetMask = ip.substring(index + 1);

        try
        {
            InetAddress ipAddress = InetAddress.getByName(subnetMask);
            byte[] mask = ipAddress.getAddress();

            Enumeration<NetworkInterface> nets = NetworkInterface.getNetworkInterfaces();
            for (NetworkInterface netint : java.util.Collections.list(nets))
            {
                java.util.List<java.net.InterfaceAddress> ifAddresses = netint.getInterfaceAddresses();
                for (java.net.InterfaceAddress ifAddress : ifAddresses) {
                    if (ifAddress.getAddress() instanceof Inet4Address) {
                        byte[] addr = ifAddress.getAddress().getAddress();
                        for (int j = 0; j < addr.length; j++) addr[j] = (byte)((int)addr[j] & (int)mask[j]);

                        String Subnet = InetAddress.getByAddress(addr).toString();
                        index = Subnet.indexOf('/');
                        if (index >= 0) Subnet = Subnet.substring(index+1);

                        if (Subnet.equals(subnetIp))
                        {
                            // split "hostname/127.0.0.1/8 [0.255.255.255]"
                            String s[] = ifAddress.toString().split("/");
                            return s[1];
                        }
                    }
                }
            }
        }
        catch (Exception e)
        {
        }
        return subnetIp;
    }

}

File: C#/Ops/Domain.cs
///////////////////////////////////////////////////////////
//  Domain.cs
//  Implementation of the Class Domain
//  Created on:      12-nov-2011 09:25:30
//  Author:
///////////////////////////////////////////////////////////

using System.Collections.Generic;

namespace Ops 
{
	public class Domain : OPSObject 
    {
		private string domainAddress = "";
		private string domainID = "";
		private string localInterface = "0.0.0.0";
		protected List<Topic> topics = new List<Topic>();
        private int timeToLive = 1;
        private int inSocketBufferSize = 16000000;
        private int outSocketBufferSize = 16000000;
		private int metaDataMcPort = 9494;

        public Domain()
        {
            AppendType("Domain");
        }
        

        public Topic GetTopic(string name)
        {
            foreach (Topic topic in topics)
            {
                if (topic.GetName().Equals(name))
                {
                    if (topic.GetDomainAddress().Equals(""))
                    {
                        topic.SetDomainAddress(domainAddress);
                    }
                    return topic;
                }
            }
            return null;
        }

        public override void Serialize(IArchiverInOut archive)
        {
            // NOTE. Keep this in sync with the C++ version, so it in theory is possible to send these as objects.
            // We need to serialize fields in the same order as C++.
            //OPSObject::serialize(archiver);
            base.Serialize(archive);

            //archiver->inout(std::string("domainID"), domainID);
            //archiver->inout<Topic>(std::string("topics"), topics);
            //archiver->inout(std::string("domainAddress"), domainAddress);
            //archiver->inout(std::string("localInterface"), localInterface);
            domainID = archive.Inout("domainID", domainID);
            topics = (List<Topic>)archive.InoutSerializableList("topics", topics);
            domainAddress = archive.Inout("domainAddress", domainAddress);
            localInterface = archive.Inout("localInterface", localInterface);
            
            //archiver->inout(std::string("timeToLive"), timeToLive);
            //archiver->inout(std::string("inSocketBufferSize"), inSocketBufferSize);
            //archiver->inout(std::string("outSocketBufferSize"), outSocketBufferSize);
            //archiver->inout(std::string("metaDataMcPort"), metaDataMcPort);
            timeToLive = archive.Inout("timeToLive", timeToLive);
    		inSocketBufferSize = archive.Inout("inSocketBufferSize", inSocketBufferSize);
	    	outSocketBufferSize = archive.Inout("outSocketBufferSize", outSocketBufferSize);
            metaDataMcPort = archive.Inout("metaDataMcPort", metaDataMcPort);
        }

        public List<Topic> GetTopics() 
        {
            foreach (Topic topic in topics) 
            {
                if (topic.GetDomainAddress().Equals("")) 
                {
                    topic.SetDomainAddress(domainAddress);
                }
            }
            return topics;
        }

         public bool ExistTopic(string name) 
         {
            foreach (Topic topic in topics) 
            {
                if (topic.GetName().Equals(name)) 
                {
                    return true;
                }
            }
            return false;
        }

        public string GetDomainAddress() 
        {
            return this.domainAddress;
        }

        public string GetDomainID() 
        {
            return this.domainID;
        }

        public string GetLocalInterface() 
        {
            return this.localInterface;
        }

        public int GetMetaDataMcPort()
        {
            return this.metaDataMcPort;
        }

        public int GetInSocketBufferSize()
        {
            return this.inSocketBufferSize;
        }

        public int GetOutSocketBufferSize()
        {
            return this.outSocketBufferSize;
        }

        public int getTimeToLive()
        {
            return this.timeToLive;
        }

        public void SetDomainAddress(string domainAddress) 
        {
            this.domainAddress = domainAddress;
        }

        public void SetDomainID(string domainID) 
        {
            this.domainID = domainID;
        }

        public void SetLocalInterface(string localInterface) 
        {
            this.localInterface = localInterface;
        }

        // If argument contains a "/" we assume it is on the form:  subnet-address/subnet-mask
        // In that case we loop over all interfaces and take the first one that matches
        // i.e. the one whos interface address is on the subnet
        public static string DoSubnetTranslation(string ip)
        {
            int index = ip.IndexOf('/');
            if (index < 0) return ip;

            string subnetIp = ip.Substring(0, index);
            string subnetMask = ip.Substring(index + 1);

            System.Net.NetworkInformation.IPGlobalProperties gp = System.Net.NetworkInformation.IPGlobalProperties.GetIPGlobalProperties();
            System.Net.NetworkInformation.UnicastIPAddressInformationCollection x = gp.GetUnicastAddresses();
            for (int i = 0; i < x.Count; i++)
            {
                if (x[i].Address.AddressFamily == System.Net.Sockets.AddressFamily.InterNetwork) //IPV4
                {
                    byte[] addr = x[i].Address.GetAddressBytes();
                    byte[] mask = x[i].IPv4Mask.GetAddressBytes();
                    for (int j = 0; j < addr.Length; j++) addr[j] = (byte)((int)addr[j] & (int)mask[j]);
                    string Subnet = new System.Net.IPAddress(addr).ToString();

                    if (Subnet.Equals(subnetIp))
                    {
                        return x[i].Address.ToString();
                    }
                }
            }
            return subnetIp;
        }

	}

}